saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+                    "scheduled": job.scheduled or seconds(now()),
+                },
+            )
+
+            if not await cursor.fetchone():
+                return None
+
+            await cursor.execute(f"NOTIFY \"{ENQUEUE_CHANNEL}\", '{job.key}'")
+
+        logger.info("Enqueuing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+        return job
+
+    async def get_abort_errors(self, jobs: Iterable[Job]) -> list[bytes | None]:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                SELECT key, job->'error'
+                FROM {jobs_table}
+                WHERE key = ANY(%(keys)s)
+                """
+                ).format(jobs_table=self.jobs_table),
+                {"keys": [job.key for job in jobs]},
+            )
+            results: dict[str, str | None] = dict(await cursor.fetchmany())
+        errors = []
+        for job in jobs:
+            error = results.get(job.key)
+            errors.append(error.encode("utf-8") if error else None)
+        return errors
+
+    async def finish_abort(self, job: Job) -> None:
+        await job.finish(Status.ABORTED, error=job.error)
+
+    async def dequeue_timer(self, poll_interval: int) -> None:
+        """Wakes up a single dequeue task every `poll_interval` seconds."""
+        while True:
+            async with self.cond:
+                self.cond.notify(1)
+            await asyncio.sleep(poll_interval)
+
+    async def listen_for_enqueues(self, timeout: float | None = None) -> None:
+        """Wakes up a single dequeue task when a Postgres enqueue notification is received."""
+        async with self.pool.connection() as conn:
+            await conn.execute(f'LISTEN "{ENQUEUE_CHANNEL}"')
+            await conn.commit()
+            gen = conn.notifies(timeout=timeout)
+            async for _ in gen:
+                async with self.cond:
+                    self.cond.notify(1)
+
+    async def _dequeue(self) -> list[Job]:
+        if not self.waiting:
+            return []
+        jobs = []
+        async with self._get_connection() as conn, conn.cursor() as cursor, conn.transaction():
+            await cursor.execute(
+                SQL(
+                    """
+                WITH locked_job AS (
+                  SELECT key, lock_key
+                  FROM {jobs_table}
+                  WHERE status = 'queued'
+                    AND queue = %(queue)s
+                    AND %(now)s >= scheduled
+                  ORDER BY scheduled
+                  LIMIT %(limit)s
+                  FOR UPDATE SKIP LOCKED
+                )
+                UPDATE {jobs_table} SET status = 'active'
+                FROM locked_job
+                WHERE {jobs_table}.key = locked_job.key AND pg_try_advisory_lock(locked_job.lock_key)
+                RETURNING job
+                """
+                ).format(jobs_table=self.jobs_table),
```

Comment:
why is jobs table passed in with format and not the params below? we shouldn't use old style over f-strings, that's not what i meant

Reply:
This is calling format on a psycopg.sql.SQL object, not on a string. This is the psycopg way to dynamically set tables and fields for a query and prevent SQL injection. We can't bind tables and fields in queries like other query values with placeholders like with the params below.

Reply:
great